### PR TITLE
Fix Sending Message

### PR DESCRIPTION
**Pull request by Simplemart17**
# Description
- This PR fixes the issue with sending and receiving messages in the chat area
- The chat messages are reordered to display the latest message last in the chat area to follow the conversation easily
- The logout button was moved next to the user badge avatar as dropdown
Screenshot from 2021-05-19 03-02-41

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -19,3 +19,5 @@ yarn-debug.log*
 yarn-error.log*
 
 server/.env
+
+.history
```
Comment (RayK3) [inline .gitignore]: Not sure why, but it appears that this is causing merge conflicts, so make sure that GitHub says it is able to merge when you go through and address these comments! :smile:

File: client/src/components/Sidebar/CurrentUser.js [hunk 2]
```diff
@@ -1,57 +1,80 @@
...
+          classes={{ root: classes.ellipsis }}
+          onClick={handleClick}
+        />
+        <Menu
+          id="logout-menu"
+          anchorEl={anchorEl}
+          keepMounted
+          open={Boolean(anchorEl)}
+          onClose={handleClose}
+          style={{ top: 30, left: -30 }}
```
Comment (RayK3) [inline client/src/components/Sidebar/CurrentUser.js]: Quick fix here: if you are already using styles via `makeStyles` above, you shouldn't use inline styles as well! It's great for quick testing, but when you have it figured out, put the styles back into the classes above!

File: client/src/store/utils/thunkCreators.js [hunk 3]
```diff
@@ -93,9 +93,9 @@ const sendMessage = (data, body) => {
 
 // message format to send: {recipientId, text, conversationId}
 // conversationId will be set to null if its a brand new conversation
-export const postMessage = (body) => (dispatch) => {
+export const postMessage = (body) => async (dispatch) => {
   try {
-    const data = saveMessage(body);
+    const data = await saveMessage(body);
```
Comment (RayK3) [inline client/src/store/utils/thunkCreators.js:86]: Nice job finding/fixing this bug

File: server/routes/api/conversations.js [hunk 4]
```diff
@@ -20,9 +20,9 @@ router.get("/", async (req, res, next) => {
         },
       },
       attributes: ["id"],
-      order: [[Message, "createdAt", "DESC"]],
+      order: [[Message, "createdAt", "ASC"]],
       include: [
-        { model: Message, order: ["createdAt", "DESC"] },
+        { model: Message, order: ["createdAt", "ASC"] },
```
Comment (RayK3) [inline server/routes/api/conversations.js]: Nice job finding this as well 👍

**Review by RayK3 — COMMENTED** [review]
So there were three bugs in total and you found two of them, namely the async/await, and message order bugs. There is also a security flaw where any user can send a message to any conversation. How would you go about fixing this?

**Comment by Simplemart17** [comment]
> So there were three bugs in total and you found two of them, namely the async/await, and message order bugs. There is also a security flaw where any user can send a message to any conversation. How would you go about fixing this?

I noticed sending a message to a conversation require a `conversationId`,  but for new conversation, there a process of creating the conversation first before sending the message. However, checking the database to search for `conversationId` could mean that any user could retrieve a `conversationId` and send message to the said conversation once the `senderId` and the `recipientId` is known. I was able to fix this.

File: server/routes/api/conversations.js (hunk 4 above)
Comment (ThinkTink) [reply server/routes/api/conversations.js]: So actually, one consequence of fixing it like this is that now the conversations are in the wrong order - we want conversations to be listed in the order of which received a message most recently

File: server/routes/api/messages.js [hunk 5]
```diff
@@ -8,36 +8,30 @@ router.post("/", async (req, res, next) => {
...
     }
+
     const senderId = req.user.id;
     const { recipientId, text, conversationId, sender } = req.body;
 
     // if we already know conversation id, we can save time and just add it to message and return
     if (conversationId) {
       const message = await Message.create({ senderId, text, conversationId });
       return res.json({ message, sender });
-    }
```
Comment (ThinkTink) [inline server/routes/api/messages.js:18]: This isn't quite right either, it's the assumption right above this that any conversationId can have a message sent to that's the issue!

**Review by ThinkTink — APPROVED** [review]
A few more tweaks, then this one is good to merge!

**Pull request merged**